src/Core/tests/gtest_DecimalFunctions.cpp

```diff
@@ -170,3 +207,27 @@ INSTANTIATE_TEST_SUITE_P(Basic,
         }
     })
 );
+
+INSTANTIATE_TEST_SUITE_P(Basic,
+    DecimalUtilsSplitAndCombineForDateTime64Test,
+    ::testing::ValuesIn(std::initializer_list<DecimalUtilsSplitAndCombineTestParam>{
+        {
+            "Negative timestamp 1965-12-12 12:12:12.123 UTC",
+            DateTime64(-127943267877),
```

Comment:
I encourage you to check if these tests don't work on master.

Reply:
No. It's not

``` sql
:) select toDateTime64('1965-12-12 12:12:12.123', 3, 'UTC') as a,  cast(a as Decimal64(3))
┌───────────────────────a─┬─CAST(toDateTime64('1965-12-12 12:12:12.123', 3, 'UTC'), 'Decimal64(3)')─┐
│ 1965-12-12 12:12:12.123 │                                                          -127914468.123 │
└─────────────────────────┴─────────────────────────────────────────────────────────────────────────┘
```

Reply:
unit tests will pass if you apply the following diff:
``` bash
--- a/src/Core/tests/gtest_DecimalFunctions.cpp
+++ b/src/Core/tests/gtest_DecimalFunctions.cpp
@@ -216,8 +216,8 @@ INSTANTIATE_TEST_SUITE_P(Basic,
             DateTime64(-127943267877),
             3,
             {
-                -127943268,
-                123
+                -127943267,
+                877
             }
         },
```
and for me, everything looks like your fix fixes the problem because only one function ( `decimalFromComponentsWithMultiplier`) is getting called during the query execution.

Reply:
You mean we must keep the sign of `whole` and `fractional` same in `DecimalComponents`  ?